client/verta/verta/pipeline/_pipeline_graph.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+
+from typing import Any, Dict, List
+
+from verta._internal_utils._utils import Configuration, Connection
+from ._pipeline_step import PipelineStep
+
+
+class PipelineGraph:
+    """
+    A collection of PipelineSteps to be run as a single inference pipeline.
+
+    Parameters
+    ----------
+    steps : list of :class:`~verta.pipeline.PipelineStep`
+        List of all possible steps of the pipeline. Ordering of steps in the pipeline
+        itself is determined by the predecessors provided to each step, thus ordering
+        of this list is irrelevant.
+
+    Attributes
+    ----------
+        steps: list of :class:`~verta.deployment.PipelineStep`
+            List of PipelineSteps comprising all possible steps in the PiplineGraph.
+    """
+
+    def __init__(self, steps: List[PipelineStep]):
+        self._steps = steps
+
+    def __repr__(self):
+        return "\n".join((f"\nPipelineGraph steps:\n{self._format_steps()}",))
+
+    def _format_steps(self):
+        """Format steps for improved readability in __repr__() function."""
+        return "\n".join([str(s) for s in self._steps])
+
+    @property
+    def steps(self):
+        return self._steps
```

Comment:
I'm not sure if/how we can prevent users from attempting such mutations, but I think we can catch it more effectively.  Take a look at the current solution and tell me what you think.  In summary:

* Validation logic is now contained in a separate function, so it can be called during __init__ and upon setting or accessing the class attributes.
* The PipelineGraph's init method now calls the PipelineStep's `predecessors` attribute for each step, so any inappropriate mutation of `predecessors` would raise an exception the moment they try to use those steps to create a graph.  Additionally, RegisteredPipeline's init method calls the PipelineGraph's `steps` attribute, so any inappropriately mutation of `steps` would throw an exception if they try to use that graph to create a pipeline.